pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,446 @@
+
+// onNodeLatencyMonitorAdd is the event handler for adding NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) onNodeLatencyMonitorAdd(obj interface{}) {
+	nlm := obj.(*v1alpha1.NodeLatencyMonitor)
+	klog.V(4).InfoS("NodeLatencyMonitor added", "NodeLatencyMonitor", klog.KObj(nlm))
+
+	m.updateLatencyConfig(nlm)
+}
+
+// onNodeLatencyMonitorUpdate is the event handler for updating NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) onNodeLatencyMonitorUpdate(oldObj, newObj interface{}) {
+	oldNLM := oldObj.(*v1alpha1.NodeLatencyMonitor)
+	newNLM := newObj.(*v1alpha1.NodeLatencyMonitor)
+	klog.V(4).InfoS("NodeLatencyMonitor updated", "NodeLatencyMonitor", klog.KObj(newNLM))
+
+	if oldNLM.GetGeneration() == newNLM.GetGeneration() {
+		return
+	}
+
+	m.updateLatencyConfig(newNLM)
+}
+
+// updateLatencyConfig updates the latency config based on the NodeLatencyMonitor CRD.
+func (m *NodeLatencyMonitor) updateLatencyConfig(nlm *v1alpha1.NodeLatencyMonitor) {
+	pingInterval := time.Duration(nlm.Spec.PingIntervalSeconds) * time.Second
+
+	m.latencyConfig = &LatencyConfig{
+		Enable:   true,
+		Interval: pingInterval,
+	}
+
+	m.latencyConfigChanged <- struct{}{}
+}
+
+// onNodeLatencyMonitorDelete is the event handler for deleting NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) onNodeLatencyMonitorDelete(obj interface{}) {
+	m.latencyConfig = &LatencyConfig{Enable: false}
+	klog.V(4).InfoS("NodeLatencyMonitor deleted")
+
+	m.latencyConfigChanged <- struct{}{}
+}
+
+// sendPing sends an ICMP message to the target IP address.
+func (m *NodeLatencyMonitor) sendPing(socket net.PacketConn, addr net.IP) error {
+	var requestType icmp.Type
+
+	ip := &net.IPAddr{IP: addr}
+
+	if addr.To4() == nil {
+		requestType = ipv6.ICMPTypeEchoRequest
+	} else {
+		requestType = ipv4.ICMPTypeEcho
+	}
+
+	timeStart := time.Now()
+	seqID := getICMPSeq()
+	body := &icmp.Echo{
+		ID:   icmpEchoID,
+		Seq:  int(seqID),
+		Data: []byte(timeStart.Format(time.RFC3339Nano)),
+	}
+	msg := icmp.Message{
+		Type: requestType,
+		Code: 0,
+		Body: body,
+	}
+	klog.V(4).InfoS("Sending ICMP message", "IP", ip, "SeqID", seqID, "body", body)
+
+	// Serialize the ICMP message
+	msgBytes, err := msg.Marshal(nil)
+	if err != nil {
+		return err
+	}
+
+	// Send the ICMP message
+	_, err = socket.WriteTo(msgBytes, ip)
+	if err != nil {
+		return err
+	}
+
+	// Create or update the latency store
+	mutator := func(entry *NodeIPLatencyEntry) {
+		entry.LastSendTime = timeStart
+	}
+	m.latencyStore.SetNodeIPLatencyEntry(addr.String(), mutator)
+
+	return nil
+}
+
+// recvPing receives an ICMP message from the target IP address.
+func (m *NodeLatencyMonitor) recvPing(socket net.PacketConn, isIPv4 bool) {
+	// We only expect small packets, if we receive a larger packet, we will drop the extra data.
+	readBuffer := make([]byte, 128)
+	for {
+		// If the socket is closed, it will return
+		n, peer, err := socket.ReadFrom(readBuffer)
+		if err != nil {
+			klog.ErrorS(err, "Failed to read ICMP message")
```

Comment:
Add the following comment above this line:
```go
// When the socket is closed in the Run method, this error will be logged, which is not ideal.
// In the future, we may try setting a ReadDeadline on the socket before each ReadFrom and using
// a channel to signal that the loop should terminate.
```

Reply:
Thanks, I have changed it.